### ISSUE TITLE
codegen creates /var/tmp/starcounter with exclusive permissions

### ISSUE THREAD

**Issue by linkdata** [Starcounter/Starcounter#3939]
On Linux, creating a directory as a given user will by default not allow other users to access it. Meaning that if you run Bluestar as one user, no other user can run it as they'll get permission denied errors.

Rather than creating `/var/tmp/starcounter`, either place the files in the current users home directory `~/starcounter/tmp` or construct a path using the current username (`/var/tmp/$USER/starcounter`).

**Comment by mtornwall** [comment]
Generated object files are machine-dependent and should go in `/var/tmp/...`. `/home` could, for all we know, be mounted via NFS and shared by machines of multiple architectures.

**Comment by linkdata** [comment]
> could ... be mounted via NFS and shared by machines of multiple architectures.

Sure, but Bluestar only supports Intel, so that would not be an issue. Anyway, the point of *this* issue is, don't use one path for every user and then have permissions that deny all but one access.

**Comment by mtornwall** [comment]
> Sure, but Bluestar only supports Intel, so that would not be an issue.

Good point, but it could still be an issue if you have two machines sharing `/home` via NFS and having mutually incompatible versions of Starcounter installed.

**Comment by linkdata** [comment]
> Good point, but it could still be an issue if you have two machines sharing /home via NFS and having mutually incompatible versions of Starcounter installed.

Please, stay on the topic of the issue. If you want, feel free to open a separate issue on sharing generated files from a Starcounter installation between multiple machines.

**Comment by mtornwall** [comment]
Here you are proposing two possible ways of resolving the issue:

> Rather than creating /var/tmp/starcounter, either place the files in the current users home directory ~/starcounter/tmp or construct a path using the current username (/var/tmp/$USER/starcounter).

One of them I'd like to see implemented; the other not. I don't think voicing concerns over one of the proposals amounts to going off topic.

Nevertheless, I have nothing further to add to this thread.

**Comment by linkdata** [comment]
> One of them I'd like to see implemented; the other not. I don't think voicing concerns over one of the proposals amounts to going off topic.

Good point. I will amend the OP.

**Comment by dan31** [comment]
CC @alemoi

**Comment by alemoi** [comment]
After the talk on Friday current temp directory is /tmp/starcounter/[user]

**Comment by mtornwall** [comment]
@alemoi 

> After the talk on Friday current temp directory is /tmp/starcounter/[user]

Why the switch to `/tmp` instead of `/var/tmp`? To quote the [Filesystem Hierarchy Standard](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s18.html):

> Programs must not assume that any files or directories in /tmp are preserved **between invocations of the program**.

Unlike `/var/tmp`, which is explicitly required to be preserved between reboots.

**Comment by alemoi** [comment]
Just consulted with @linkdata and he had some good points that most modern systems have /tmp as ram disk and Linux machine is not restarted that often. But "Programs must not assume that any files or directories in /tmp are preserved between invocations of the program." seems serious drawback. @linkdata do you have any points? Thanks!

**Comment by linkdata** [comment]
It's a drawback if we need them to persist across reboots. It's a feature if we don't (because then we won't have to do housekeeping for them ourselves). Also, /tmp is usually in RAM.

**Comment by alemoi** [comment]
> It's a drawback if we need them to persist across reboots. It's a feature if we don't (because then we won't have to do housekeeping for them ourselves). Also, /tmp is usually in RAM.

Thanks for the feedback. Though @mtornwall points that /tmp can be cleaned even between process invocations, that's what got me confused.

**Comment by linkdata** [comment]
> Thanks for the feedback. Though @mtornwall points that /tmp can be cleaned even between process invocations, that's what got me confused.

Oh, I totally missed that! Thanks for pointing it out, @mtornwall. I don't think it changes anything as it applies to codegen though. I mean, initially we wanted it all in RAM for the duration of the codehost process only. 

Also, while the standard does say that, I can't think of any Linux distribution that actually cleans out /tmp that aggressively. As long as we're fine with a file not being there and can rebuild it when needed, it shouldn't impact us.

I'm fine with it either location.

**Comment by mtornwall** [comment]
> It's a feature if we don't (because then we won't have to do housekeeping for them ourselves).

Good point.

> Also, while the standard does say that, I can't think of any Linux distribution that actually cleans out /tmp that aggressively.

No, this is probably a purely academic problem.

> I'm fine with it either location.

Me too.
